### PR TITLE
vfio-pci: Add a nic when boot guest with pf/vf

### PR DESCRIPTION
**Pull request by huangyum**
Add a nic when boot guest with pf/vf, so guest can get an IP address, multi_pf_boot and vf_boot cases can pass.

Signed-off-by: Yumei Huang <yuhuang@redhat.com>
ID: 1468525, 1468528

Depens on https://github.com/avocado-framework/avocado-vt/pull/1087

### CODE REVIEW

**Comment by huangyum** [comment]
Test on a host with XL710 and BCM57810 card: 
(1/4) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.pf_assignable.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.device_assignment.vfio-pci.multi_pf_boot.82576: SKIP
 (2/4) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.pf_assignable.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.device_assignment.vfio-pci.multi_pf_boot.XL710: PASS (176.57 s)
 (3/4) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.pf_assignable.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.device_assignment.vfio-pci.multi_pf_boot.82599: SKIP
 (4/4) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.pf_assignable.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.io-github-autotest-qemu.device_assignment.vfio-pci.multi_pf_boot.BCM57810: PASS (75.03 s)

File: qemu/tests/cfg/sr_iov.cfg [hunk 1]
```diff
@@ -1,12 +1,27 @@
...
         nic_interface_filter = "(eth\d*)\s*Link encap:Ethernet\s*HWaddr .\w+:\w+:\w+:\w+:\w+:\w+\s*inet addr:\d+.\d+.\d+.\d+"
     variants:
         - vf_boot:
             type = boot
-            pci_assignable_nic1 = vf
+            nics = "nic1 nic2"
+            pci_assignable_nic1 = no
+            nic_model_nic1 = virtio
+            pci_assignable_nic2 = vf
+            host_setup_flag = 3
```
Comment (yanan-fu) [inline qemu/tests/cfg/sr_iov.cfg]: Move out of the variants is better.

File: qemu/tests/cfg/device_assignment.cfg [hunk 2]
```diff
@@ -1,13 +1,26 @@
 - device_assignment:
     only pf_assignable
     virt_test_type = qemu
+    pre_command = "modprobe vfio-pci"
```
Comment (yanan-fu) [inline qemu/tests/cfg/device_assignment.cfg]: In test_setup.py, function sr_iov_setup() will check this, no need to add it here.
If sr_iov_setup() have bug  for this operation, bug fix is enough, i think.

And If can not reuse sr_iov_setup at last, move this to guest_hw.cfg for "pf_assignable" and "vf_assignable" is better.

File: qemu/tests/cfg/sr_iov.cfg [hunk 3]
```diff
@@ -1,12 +1,27 @@
 - sr-iov:
     only vf_assignable
     nic_interface_filter = "(em\d+).*inet .\d+.\d+.\d+.\d+.*?ether .\w+:\w+:\w+:\w+:\w+:\w+"
+    pre_command = "modprobe vfio-pci"
```
Comment (yanan-fu) [inline qemu/tests/cfg/sr_iov.cfg]: ditto

File: qemu/tests/cfg/device_assignment.cfg (hunk 2 above)
Comment (huangyum) [inline qemu/tests/cfg/device_assignment.cfg]: sr_iov_setup() will not be called for device_assignment.  I will replace it in guest-hw.cfg. Thanks.

Comment (yanan-fu) [inline qemu/tests/cfg/device_assignment.cfg]: why it was not called ?   From my checking,  it will be called if host_set_flag was set to setup env.

Comment (huangyum) [inline qemu/tests/cfg/device_assignment.cfg]: Yes, but we don't need to set the flag for device_assignment. sr_iov_setup() is used to set vf for pf.  No vf is needed in this case.

Comment (yanan-fu) [inline qemu/tests/cfg/device_assignment.cfg]: we need to check interrupt remapping and vfio-pci module for all device assignment  test, no matter pf or vf.  Only one pre_command  to add vfio-pci  module is not enough.

Comment (huangyum) [inline qemu/tests/cfg/device_assignment.cfg]: Updated, please help review again. Thanks!

File: qemu/tests/cfg/device_assignment.cfg [hunk 4]
```diff
@@ -7,7 +7,20 @@
     variants:
         - multi_pf_boot:
             type = boot
-            nics = "nic1 nic2"
+            nics = "nic1 nic2 nic3"
```
Comment (suqinhuang) [inline qemu/tests/cfg/device_assignment.cfg]: does the script works? seems we have give up to develop the script about device assignment

**Comment by suqinhuang** [comment]
@CongLi @luckyh better to double check it if we can develop device assignment/sriov related cases, otherwise we will take lots of time to maintain the script

File: qemu/tests/cfg/sr_iov.cfg [hunk 5]
```diff
@@ -6,7 +6,21 @@
     variants:
         - vf_boot:
             type = boot
-            pci_assignable_nic1 = vf
+            nics = "nic1 nic2"
+            pci_assignable_nic1 = no
+            nic_model_nic1 = virtio
+            pci_assignable_nic2 = vf
+            only x86_64
```
Comment (luckyh) [inline qemu/tests/cfg/sr_iov.cfg]: IMHO, this test also support, at least, ppc.

Comment (huangyum) [inline qemu/tests/cfg/sr_iov.cfg]: I put it here because ppc team doesn't test the following four cards. Maybe we could submit another patch once ppc test is needed. Does it make any sense to you ?

Comment (luckyh) [inline qemu/tests/cfg/sr_iov.cfg]: @balamuruhans I guess you won't be happy with this change, will you?

**Comment by zhguo01** [comment]
LGTM

Test on a host with XL710 and 82599 sr-iov capable nics, results:
vf test:
(1/4) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.vf_assignable.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.zhguo.sr-iov.vfio-pci.vf_boot.82576: SKIP
(2/4) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.vf_assignable.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.zhguo.sr-iov.vfio-pci.vf_boot.XL710: PASS (58.38 s)
(3/4) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.vf_assignable.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.zhguo.sr-iov.vfio-pci.vf_boot.82599: PASS (75.02 s)
(4/4) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.vf_assignable.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.zhguo.sr-iov.vfio-pci.vf_boot.BCM57810: SKIP

pf test:

 (1/4) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.pf_assignable.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.zhguo.device_assignment.vfio-pci.multi_pf_boot.82576: SKIP
 (2/4) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.pf_assignable.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.zhguo.device_assignment.vfio-pci.multi_pf_boot.XL710: PASS (148.11 
s)
 (3/4) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.pf_assignable.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.zhguo.device_assignment.vfio-pci.multi_pf_boot.82599: PASS (73.94 s
)
 (4/4) smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.pf_assignable.qcow2.virtio_blk.up.virtio_net.RHEL.7.4.x86_64.zhguo.device_assignment.vfio-pci.multi_pf_boot.BCM57810: SKIP

Acked-by: Zhiyi Guo zhguo@redhat.com

File: qemu/tests/cfg/device_assignment.cfg [hunk 6]
```diff
@@ -7,7 +7,19 @@
...
+            driver_option = 0
+            variants:
+                - 82576:
+                    pf_filter_re = 82576
+                - XL710:
+                    pf_filter_re = XL710
+                - 82599:
+                    pf_filter_re = 82599
+                - BCM57810:
+                    pf_filter_re = BCM57810
```
Comment (balamuruhans) [inline qemu/tests/cfg/device_assignment.cfg]: @luckyh, also I doubt will this test be generic, for every card we need to add it here ?

Comment (luckyh) [inline qemu/tests/cfg/device_assignment.cfg]: @balamuruhans have the same concern with you, however, I can't say which is more user friendly. To developers, it's indeed annoying to add variant for every new card. To testers, they do want less manual configuration.

File: qemu/tests/cfg/sr_iov.cfg (hunk 5 above)
Comment (huangyum) [inline qemu/tests/cfg/sr_iov.cfg]: I will remove it from here. Thanks!

File: qemu/tests/cfg/device_assignment.cfg (hunk 6 above)
Comment (yanan-fu) [inline qemu/tests/cfg/device_assignment.cfg]: @balamuruhans  i think it is ok here. The old design list only one nic, and request tester to change "pf_filter_re" according to the test environment.  The new design list more nics, so i think it can lighten the workload.
If no nic match with the test environment, tester can change it again, this is same with the old design. what do you think ?

Comment (huangyum) [inline qemu/tests/cfg/device_assignment.cfg]: Thanks all for your review. 
With this patch, testers don't need to configure "pf_filter_re" by manual and  test the cards one by one, which could save much effort. 
I agree with @luckyh , it may be annoying to developers to add variant for every new card.  
I don't know if there is a tradeoff between testers and developers. If you have any better idea, you are welcome to bring up. Thanks!

Comment (balamuruhans) [inline qemu/tests/cfg/device_assignment.cfg]: @luckyh @yanan-fu , I would prefer to have variants in test script specific cfg (like sriov.cfg) with different test scenarios instead of same test with different cards. I agree that we need less manual configuration for which I would have the consolidated test cases in cfg which will be given to --vt-config. so the different cards can be added as variants in this cfg with the test case.

Comment (huangyum) [reply qemu/tests/cfg/device_assignment.cfg]: @balamuruhans @zhguo01 @yanan-fu @luckyh  I've moved the cards types to some other cfg. Please help review again, thanks!

File: qemu/tests/cfg/sr_iov.cfg [hunk 7]
```diff
@@ -6,7 +6,11 @@
     variants:
         - vf_boot:
             type = boot
-            pci_assignable_nic1 = vf
+            nics = "nic1 nic2"
+            pci_assignable_nic1 = no
+            nic_model_nic1 = virtio
+            pci_assignable_nic2 = vf
```
Comment (yanan-fu) [inline qemu/tests/cfg/sr_iov.cfg]: No need add it again, duplicate  with line 2 "only vf_assignable"

**Comment by huangyum** [comment]
Closed as the case is covered by vfio-pci.emulation_nic_pf_boot.nic_virtio.
